expression/builtin_regexp.go

```diff
@@ -1096,43 +1118,189 @@ func (re *builtinRegexpReplaceFuncSig) Clone() builtinFunc {
 	return newSig
 }
 
-func (re *builtinRegexpReplaceFuncSig) getReplacedBinStr(reg *regexp.Regexp, bexpr []byte, trimmedBexpr []byte, repl string, pos int64, occurrence int64) (string, bool, error) {
-	count := occurrence
-	repFunc := func(matchedStr []byte) []byte {
-		if occurrence == 0 {
-			return []byte(repl)
+func (re *builtinRegexpReplaceFuncSig) processEmptyStr(reg *regexp.Regexp, instructions []Instruction, pos int64, occurrence int64) ([]byte, error) {
+	if occurrence > 1 || pos != 1 {
+		return []byte(""), nil
+	}
+
+	emptyStr := make([]byte, 0)
+	res := reg.FindSubmatchIndex(emptyStr)
+	if len(res) != 0 {
+		replacedBStr := make([]byte, 0)
+		// Copy replacement
+		for _, instruction := range instructions {
+			if instruction.SubstitutionNum != -1 {
+				capturedStr, err := instruction.getCaptureGroupStr(emptyStr, res)
+				if err != nil {
+					return []byte(""), err
+				}
+				replacedBStr = append(replacedBStr, capturedStr...)
+			} else {
+				replacedBStr = append(replacedBStr, instruction.Literal...)
+			}
+		}
+		return replacedBStr, nil
+	}
+	return []byte(""), nil
+}
+
+func (re *builtinRegexpReplaceFuncSig) replaceAllMatchedBinStr(reg *regexp.Regexp, bexpr []byte, trimmedBexpr []byte, instructions []Instruction, pos int64) ([]byte, error) {
+	replacedBStr := make([]byte, 0)
+	for {
+		res := reg.FindSubmatchIndex(trimmedBexpr)
+		if len(res) == 0 {
+			break
 		}
 
-		count--
-		if count == 0 {
-			return []byte(repl)
+		replacedBStr = append(replacedBStr, trimmedBexpr[:res[0]]...) // Copy prefix
+
+		// Copy replacement
+		for _, instruction := range instructions {
```

Comment:
I think we can create a function to do the `Copy replacement` things, and let `processEmptyStr/replaceAllMatchedBinStr/replaceOneMatchedBinStr` call the function

Reply:
> 

okk